src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php

```diff
@@ -112,35 +112,51 @@ public function handle(GetResponseEvent $event)
      */
     private function attemptSwitchUser(Request $request)
     {
-        $token = $this->securityContext->getToken();
-        $originalToken = $this->getOriginalToken($token);
-
-        if (false !== $originalToken) {
-            if ($token->getUsername() === $request->get($this->usernameParameter)) {
-                return $token;
-            } else {
-                throw new \LogicException(sprintf('You are already switched to "%s" user.', $token->getUsername()));
-            }
+        if (null !== $this->logger) {
+            $this->logger->info(sprintf('Attempt to switch to user "%s"', $request->get($this->usernameParameter)));
         }
 
-        if (false === $this->accessDecisionManager->decide($token, array($this->role))) {
+        // token of the currently authenticated user
+        $sourceToken = $this->securityContext->getToken();
+
+        // check if the authenticated user has the right role to switch user
+        if (false === $this->accessDecisionManager->decide($sourceToken, array($this->role)) && false === $this->accessDecisionManager->decide($sourceToken, array('ROLE_PREVIOUS_ADMIN'))) {
             throw new AccessDeniedException();
         }
 
-        $username = $request->get($this->usernameParameter);
-
-        if (null !== $this->logger) {
-            $this->logger->info(sprintf('Attempt to switch to user "%s"', $username));
+        // user is attempting to switch to his own username
+        if ($sourceToken->getUsername() === $request->get($this->usernameParameter)) {
+            // throw new \LogicException() ?
+            return $sourceToken;
```

Comment:
I'll check if a test exists.
Anyway if we want to be strict, this should throw an exception because switching to the same user is a weird behavior; I don't know, maybe I'd leave this as is (returning $sourceToken) because nobody would be interested in catching the LogicException thrown